### ISSUE TITLE
using BigMPI for chunks greater than 6GB

### ISSUE THREAD

**Issue by Aazmp** [open-mpi/ompi#5069]
Thank you for taking the time to submit an issue!

## Background information
with https://github.com/jeffhammond/BigMPI 
Trying to transfer huge blocks (> 6gb)
irrelevant of pml used data is getting cut

### What version of Open MPI are you using? (e.g., v1.10.3, v2.1.0, git branch name and hash, etc.)
openmpi-3.1.0rc3 from open-mpi.org

### Please describe the system on which you are running
-----------------------------
these are irrelevant, but testing is done on CentOS 7 with with nodes connected over tcp or located on same motherboard

## Details of the problem
BigMPI tries to create huge datatype from two parts called "chunks" and "remainmder"
first one is Multiple of INT_MAX, second is remainder

chunks are added to new type with 
``` c++
MPI_Type_vector(count, INT_MAX, INT_MAX, ...)
```
later being passed to 
``` c++
"ompi_datatype_add( pData, oldType, !!!count * bLength!!!, 0, extent )"
```
where bLength is already INT_MAX and it's being cut after multiplication (that is also int*int)

new type is created only up to 6Gb as remainder can append up to 2GB and ompi_datatype_add's param is being cut to 4gb (uint32_t). With any message greater than 6GB really transferred size is 2GB + remainder. 

For simple case of send recv this works for me 
1. ompi/datatype/ompi_datatype.h: 
``` c
  from   ompi_datatype_add( ..., uint32_t count,  ... );
  to       ompi_datatype_add( ..., size_t count,  ... );
```
2. opal/datatype/opal_datatype.h:
``` c
  from   opal_datatype_add( ..., uint32_t count,  ... );
  to       opal_datatype_add( ..., size_t count,  ... );
```
3. opal/datatype/opal_datatype_add.c:
``` c
  from   opal_datatype_add( ..., uint32_t count,  ... );
  to       opal_datatype_add( ..., size_t count,  ... );
```
4. ompi/datatype/ompi_datatype_create_vector.c
``` c++
  from  ompi_datatype_add( pData, oldType, count * bLength, 0, extent );
  to      ompi_datatype_add( pData, oldType, (size_t)count * (size_t)bLength, 0, extent );
```

**Comment by ggouaillardet** [comment]
@Aazmp thanks for the report.
would you be able to write a simple program that can evidence the issue ?
this can be a high level test (e.g. pure `MPI`) or low level (see the tests in `test/datatype`), if not, then a `BigMPI` test will be enough.

**Comment by hjelmn** [comment]
@bosilca Can this be closed? Did the fix get into v3.0.x and v3.1.x?

**Comment by bosilca** [comment]
Fixed by #5072

**Comment by bosilca** [comment]
I take it back the fix is in master but never made it into our release branches.

**Comment by jsquyres** [comment]
* PR's filed for v3.0.x and v3.1.x.
* Not an issue for v4.0.x (I assume this was merged to master before v4.0.x branched).
* Executive decision: not worth it for v2.1.x.

**Comment by awlauria** [comment]
Closing as all PR's merged.
